src/python_testing/TC_RR_1_1.py

```diff
@@ -378,6 +433,168 @@ async def fill_user_label_list(self, dev_ctrl, target_node_id):
+                asserts.assert_equal(group_keys[fabric_idx][group_key_list_idx].epochStartTime1,
+                                     key_set.groupKeySet.epochStartTime1)
+                asserts.assert_equal(group_keys[fabric_idx][group_key_list_idx].epochStartTime2,
+                                     key_set.groupKeySet.epochStartTime2)
+                asserts.assert_equal(NullValue, key_set.groupKeySet.epochKey0, "Unexpected key response from read.")
+                asserts.assert_equal(NullValue, key_set.groupKeySet.epochKey1, "Unexpected key response from read.")
+                asserts.assert_equal(NullValue, key_set.groupKeySet.epochKey2, "Unexpected key response from read.")
+
+        return group_keys
+
+    async def fill_and_validate_group_key_map(self,
+                                              fabrics: int,
+                                              clients: List[Any],
+                                              group_key_map: List[Dict[int, int]]) -> None:
+        # Step 13: Write and verify indicated_max_groups_per_fabric group/key mappings for all fabrics.
+        mapping_structs: List[List[Clusters.GroupKeyManagement.Structs.GroupKeyMapStruct]] = [[] for _ in range(fabrics)]
+        for fabric_idx in range(fabrics):
+            client: Any = clients[fabric_idx]
+            fabric_idx_matter: int = fabric_idx + 1
+
+            for group in group_key_map[fabric_idx]:
+                mapping_structs[fabric_idx].append(Clusters.GroupKeyManagement.Structs.GroupKeyMapStruct(groupId=group,
+                                                                                                         groupKeySetID=group_key_map[fabric_idx][group],
+                                                                                                         fabricIndex=fabric_idx_matter))
+
+            logging.info("Step 13: Setting group key map on fabric %d" % (fabric_idx_matter))
+            await client.WriteAttribute(self.dut_node_id, [(0, Clusters.GroupKeyManagement.Attributes.GroupKeyMap(mapping_structs[fabric_idx]))])
+
+        # Step 13 verification: After all the group key maps were written, read all the information back.
+        for fabric_idx in range(fabrics):
+            client: Any = clients[fabric_idx]
+            fabric_idx_matter: int = fabric_idx + 1
+
+            logging.info("Step 13: Reading group key map on fabric %d" % (fabric_idx_matter))
+            group_key_map_readback = await self.read_single_attribute(client, node_id=self.dut_node_id, endpoint=0, attribute=Clusters.GroupKeyManagement.Attributes.GroupKeyMap)
+
+            found_entry: int = 0
+            for read_entry in group_key_map_readback:
+                if read_entry.fabricIndex != fabric_idx_matter:
+                    continue
+
+                written_entry = next(entry for entry in mapping_structs[fabric_idx] if entry.groupId == read_entry.groupId)
```

Comment:
Not sure how `next` works here

Reply:
Yes, next will always return the first element produced by the generator, in this case the first mapping with a matching groupId. Not the most efficient to call this each time, but I'm not assuming an ordering to the entries.